src/dataworker/Dataworker.ts

```
@@ -1759,94 +1723,90 @@ export class Dataworker {
+      // This would only happen if the starting hub pool balance is below the net send amount. If it started
+      // above, then the dataworker would not purposefully send tokens out of it to fulfill the Ethereum
+      // PoolRebalanceLeaf and then return tokens to it to execute another chain's PoolRebalanceLeaf.
+      const tokenSymbol = this.clients.hubPoolClient.getTokenInfo(hubPoolChainId, l1Token)?.symbol;
+      if (currHubPoolLiquidReserves.gte(requiredNetSendAmountForL1Token)) {
         this.logger.debug({
-          at: "Dataworker#executePoolRebalanceLeaves",
-          message: `Relayer refund leaf will return enough funds to HubPool to execute PoolRebalanceLeaf, updating exchange rate for ${tokenSymbol}`,
-          updatedLiquidReserves,
-          virtualHubPoolBalance,
-          netSendAmount: leaf.netSendAmounts[idx],
-          leaf,
+          at: "Dataworker#_updateExchangeRatesBeforeExecutingNonHubChainLeaves",
+          message: `Skipping exchange rate update for ${tokenSymbol} because current liquid reserves > required netSendAmount`,
+          currHubPoolLiquidReserves,
+          requiredNetSendAmountForL1Token,
+          l1Token,
         });
-        updatedL1Tokens.add(l1Token);
+        return;
+      }
+
+      // @dev: Virtual balance = post-sync liquid reserves + any used balance.
+      const multicallInput = [
+        hubPool.interface.encodeFunctionData("sync", [l1Token]),
+        hubPool.interface.encodeFunctionData("pooledTokens", [l1Token]),
+      ];
+      const multicallOutput = await hubPool.callStatic.multicall(multicallInput);
+      const updatedPooledTokens = hubPool.interface.decodeFunctionResult("pooledTokens", multicallOutput[1]);
+      const updatedLiquidReserves = updatedPooledTokens.liquidReserves;
+
+      // If the virtual balance is still too low to execute all the pool leaves, then log an error
+      if (updatedLiquidReserves.lt(requiredNetSendAmountForL1Token)) {
+        throw new Error(
```

Similar to https://github.com/across-protocol/relayer/pull/1933/files#r1864476640, throw error rather than just log PD alert

---

RE https://github.com/across-protocol/relayer/pull/1933/files#r1872013595, should we try to execute as much as we can in order to minimise disruption to the protocol?

---

new logic to try to execute as many non hub chain leaves as possible: 41a2aae1a44d95ddc543534fa5f0888c6e3ac82a